backend/core/src/main/java/org/sonarsource/sonarlint/core/commons/SmartCancelableLoadingCache.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * SonarLint Core - Implementation
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.commons;
+
+import com.google.common.util.concurrent.MoreExecutors;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.BiFunction;
+import javax.annotation.Nullable;
+import org.sonarsource.sonarlint.core.commons.log.SonarLintLogger;
+import org.sonarsource.sonarlint.core.commons.progress.SonarLintCancelChecker;
+
+/**
+ * A cache with async computation of values, and supporting cancellation.
+ * "Smart" because when a computation is cancelled, it will return to the previous callers the result of the new computation.
+ */
+public class SmartCancelableLoadingCache<K, V> implements AutoCloseable {
+
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+
+  private final ExecutorService executorService;
+  private final String threadName;
+  private final BiFunction<K, SonarLintCancelChecker, V> valueComputer;
+  private final ConcurrentHashMap<K, ValueAndComputeFutures> cache = new ConcurrentHashMap<>();
+
+  private class ValueAndComputeFutures {
+    private final K key;
+    private CompletableFuture<V> valueFuture = new CompletableFuture<>();
```

Comment:
I always have a doubt, shouldn't this one be volatile? Or use an AtomicReference?

Reply:
Yes, I think it is a good idea